README.md
jwi - Java Wordnet Interface (support for Hadoop HDFS and Yarn)
============================

JWI 2.3.3-hadoop - fork of JWI 2.3.3 to support HDFS file system on a hadoop cluster

License
-------
See **jwi-license** file for details. All rights of original author reserved.




changelog
/******************************************************************************
 * Copyright (c) 2007-2014 Massachusetts Institute of Technology.
 * All rights reserved.
 * 
 * This program and the accompanying materials are made available under the 
 * terms of the MIT Java Wordnet Interface License which accompanies this 
 * distribution.  Please check the attached license for more details.
 *
 * To communicate praise, suggestions, constructive criticism, or feature
 * requests, write:
 *
 * Mark Alan Finlayson
 * email: markaf@alum.mit.edu
 * post: 32 Vassar St., Room 32-258, Cambridge, MA 02138 USA
 * web: http://www.mit.edu/~markaf/
 *******************************************************************************/

Changelog for edu.mit.jwi:


Ver     : Release Date  : Changes
=================================
2.3.3-haddop	: 	Jan 05, 2017 : Add support for HDFS and YARN (by mauro.pelucchi@gmail.com)
---> Fork for hadoop
2.3.3   : Feb 12, 2014  : 1. Fixed Bug 023: RAMDictionary now properly handles underspecified word id objects
----------------------------------
2.3.2   : Dec 26, 2013  : 1. Fixed Bug 022: Minor changes to stemmer to correct lemma order and correctly handle nouns and verbs of two characters or fewer
----------------------------------
2.3.1   : Dec 09, 2013  : 1. Fixed Bug 021: Factored out IStemmingRule, and added check for nouns that end with 'ss'
----------------------------------
2.3.0   : Aug 28, 2013  : 1. Added IIndexWord.getPointers() method and IndexWord implementation thereof
                          2. Added two IndexWord constructors allowing setting of pointers field
                          3. Added IndexWordParser.resolvePointer(String, POS) method and its use in the parseLine(String) method
                          4. Added Pointer.DOMAIN (symbol ';') and Pointer.MEMBER (symbol '-') so IndexWordParser.parseLine(String) doesn't throw an exception
----------------------------------
2.2.6   : Jun 25, 2013  : 1. Fixed Bug 020
---------------------------------- 
2.2.5   : Jun 17, 2013  : 1. Fixed Bug 019
                          2. IItemID now implements Serializable 
---------------------------------- 
2.2.4   : Apr 13, 2013  : 1. Fixed Bug 017; added relevant FAQ
                          2. Fixed Bug 018; tested on 3.0 database-only version; added relevant FAQ
                          3. Tested JWI on Wordnet 3.1; added Version.ver31 field 
----------------------------------        
2.2.3   : Feb 25, 2012  : 1. Fixed Bug 015
----------------------------------
2.2.2   : Nov 16, 2011  : 1. Fixed Bug 014
                          2. Changed license to CC-BY, allowing commercial use
----------------------------------
2.2.1   : Jun 15, 2011  : 1. Fixed Bug 013
----------------------------------
2.2.0   : May 17, 2011  : 1. Overhauled Javadoc and Manual
                          2. Implemented in-memory dictionary
                          3. Created IHasLifecycle and ICloseable interfaces, now implemented by IDictionary, IDataProvider, IDataSource, and ItemCache
                              a) IDataSource objects must now be opened and closed via the open() and close() methods
                              b) IDictionary.open() method now throws an IOException
                              c) Renamed DictionaryClosedException to ObjectClosedException, and moved it into the IHasLifecycle interface
                          4. Fixed Bug 012
                              a) IDataSource now implements IHasLifecycle
                              b) FileProvider now calls IDataSource.open() and .close() at appropriate places 
                          5. Condensed iterator methods in WordnetFile hierarchy
                              a) Moved iterator() and iterator(String) into WordnetFile class from subclasses
                              b) Added WordnetFile.makeIterator(ByteBuffer, String) abstract method and implemented in subclasses
                          6. Abstracted Dictionary class hierarchy a bit; moved caching functionality into one class (CachingDictionary), data source access into another (DataSourceDictionary)
                          7. Removed IDataSourceDictionary#setDataProvider and #getSources() methods, added #getTypes() method
                          8. Created the ILoadable and ILoadPolicy interfaces that allow data in WordnetFile objects to be loaded completely into memory
                          9. Changed a few method contracts, making null-handling more consistent
                          10. Renamed IWordID.getSenseNumber() to getWordNumber()
----------------------------------
2.1.5   : Dec 19, 2008  : 1. Added documentation that points from the ISenseKey interface to SenseKeyParser class
                          2. Fixed Bug 011 (WordnetStemmer.findStems(String, POS))
                          3. Make explicit that Dictionary class does no stemming, unlike Wordnet API
                          4. Iterator returned by IDictionary.getExceptionEntryIterator(POS) now returns IExceptionEntry objects
                          5. Added catch and warning for linefeed corrupted data files
---------------------------------
2.1.4   : Sep 11, 2008  : 1. Moved SenseKey.parseKey(String) to its own class, SenseKeyParser, for ease of modification
                          2. Replaced call to SenseKey.parseKey(String) in SenseLineParser with call to SenseKeyParser.parseLine(String)
                          3. Added UnknownLexFile class to allow support of Wordnet mods (such as Rion Snow's) that use non-standard lexical file numbers
                          4. Modified SenseKeyParser and DataLineParser so they use UnknownLexFile class rather than throw NullPointerException on non-standard lexical file
                          5. Fixed Bug 010 (Dictionary.getExceptionEntry(IExceptionEntryID) method)

src/edu/mit/jwi/Dictionary.java
/********************************************************************************
 * MIT Java Wordnet Interface Library (JWI) v2.3.3
 * Copyright (c) 2007-2014 Massachusetts Institute of Technology
 *
 * JWI is distributed under the terms of the Creative Commons Attribution 3.0 
 * Unported License, which means it may be freely used for all purposes, as long 
 * as proper acknowledgment is made.  See the license file included with this
 * distribution for more details.
 *******************************************************************************/

package edu.mit.jwi;

import java.io.File;
import java.net.URL;

import org.apache.hadoop.conf.Configuration;

import edu.mit.jwi.data.FileProvider;

/**
 * Basic {@code IDictionary} implementation that mounts files on disk and has
 * caching. A file URL to the directory containing the Wordnet dictionary files
 * must be provided.  This implementation has adjustable caching.
 * 
 * @author Mark A. Finlayson
 * @version 2.3.3-hadoop
 * @since JWI 1.0
 */
public class Dictionary extends CachingDictionary {

	/**
	 * Constructs a new dictionary that uses the Wordnet files located in a
	 * hdfs directory pointed to by the specified path with the configuration of Haddop
	 * 
	 * @param path
	 *            a path pointing to a hdfs directory containing the wordnet data
	 *            files on the hdsf filesystem
	 * @param conf the hadoop Configuration
	 * 
	 * @throws NullPointerException
	 *             if the specified path is <code>null</code>
	 * @since JWI 2.3.3-hadoop
	 * @author Mauro Pelucchi
	 * 
	 */
	public Dictionary(String path, Configuration conf) {
		super(new DataSourceDictionary(new FileProvider(path, conf)));
	}
	
	/**
	 * Constructs a new dictionary that uses the Wordnet files located in a
	 * directory pointed to by the specified url
	 * 
	 * @param wordnetDir
	 *            a url pointing to a directory containing the wordnet data
	 *            files on the filesystem
	 * @throws NullPointerException
	 *             if the specified url is <code>null</code>
	 * @since JWI 1.0
	 */
	public Dictionary(URL wordnetDir) {
		super(new DataSourceDictionary(new FileProvider(wordnetDir)));
	}

	/**
	 * Constructs a new dictionary that uses the Wordnet files located in a
	 * directory pointed to by the specified file
	 * 
	 * @param wordnetDir
	 *            a file pointing to a directory containing the wordnet data files on the filesystem
	 * @throws NullPointerException
	 *             if the specified file is <code>null</code>
	 * @since JWI 1.0
	 */
	public Dictionary(File wordnetDir) {
		super(new DataSourceDictionary(new FileProvider(wordnetDir)));
	}
	
}


src/edu/mit/jwi/data/DataType.java
/********************************************************************************
 * MIT Java Wordnet Interface Library (JWI) v2.3.3
 * Copyright (c) 2007-2014 Massachusetts Institute of Technology
 *
 * JWI is distributed under the terms of the Creative Commons Attribution 3.0 
 * Unported License, which means it may be freely used for all purposes, as long 
 * as proper acknowledgment is made.  See the license file included with this
 * distribution for more details.
 *******************************************************************************/

package edu.mit.jwi.data;

import java.io.File;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import org.apache.hadoop.fs.FileStatus;

import edu.mit.jwi.data.parse.DataLineParser;
import edu.mit.jwi.data.parse.ExceptionLineParser;
import edu.mit.jwi.data.parse.ILineParser;
import edu.mit.jwi.data.parse.IndexLineParser;
import edu.mit.jwi.data.parse.SenseLineParser;
import edu.mit.jwi.item.IExceptionEntryProxy;
import edu.mit.jwi.item.IIndexWord;
import edu.mit.jwi.item.ISenseEntry;
import edu.mit.jwi.item.ISynset;
import edu.mit.jwi.item.POS;

/**
 * A concrete implementation of the {@code IDataType} interface. This class
 * provides the data types necessary for Wordnet in the form of static
 * fields. It is not implemented as an {@code Enum} so that clients may add
 * their own content types by instantiating this class.
 * 
 * @param <T>
 *            the type of object for the content type
 * @author Mark A. Finlayson
 * @version 2.3.3
 * @since JWI 2.0.0
 */
public class DataType<T> implements IDataType<T> {
	
	public static final DataType<IIndexWord> 			INDEX 		= new DataType<IIndexWord>(				"Index", 	 true,	IndexLineParser.getInstance(),		"index", "idx");
	public static final DataType<ISynset> 				DATA 		= new DataType<ISynset>(			 	"Data", 	 true, 	DataLineParser.getInstance(), 		"data", "dat");
	public static final DataType<IExceptionEntryProxy> 	EXCEPTION 	= new DataType<IExceptionEntryProxy>( 	"Exception", false,	ExceptionLineParser.getInstance(),	"exception", "exc");
	public static final DataType<ISenseEntry> 			SENSE 		= new DataType<ISenseEntry>( 			"Sense", 	 false,	SenseLineParser.getInstance(),		"sense");

	// fields set on construction
    private final String name;
    private final Set<String> hints;
    private final boolean hasVersion;
    private final ILineParser<T> parser;

	/**
	 * Constructs a new data type. This constructor takes the hints as an
	 * varargs array.
	 * 
	 * @param userFriendlyName
	 *            a user-friendly name, for easy identification of this data
	 *            type; may be <code>null</code>
	 * @param hasVersion
	 *            <code>true</code> if the comment header for this data type
	 *            usually contains a version number
	 * @param parser
	 *            the line parser for transforming lines from this data type
	 *            into objects; may not be <code>null</code>
	 * @param hints
	 *            a varargs array of resource name hints for identifying the
	 *            resource that contains the data. may be <code>null</code>, but
	 *            may not contain <code>null</code>
	 * @throws NullPointerException
	 *             if the specified parser is <code>null</code>
	 * @since JWI 2.0.0
	 */
    public DataType(String userFriendlyName, boolean hasVersion, ILineParser<T> parser, String... hints) {
    	this(userFriendlyName, hasVersion, parser, (hints == null) ? null : Arrays.asList(hints));
    }
    
	/**
	 * Constructs a new data type. This constructor takes the hints as an
	 * collection.
	 * 
	 * @param userFriendlyName
	 *            a user-friendly name, for easy identification of this data
	 *            type; may be <code>null</code>
	 * @param hasVersion
	 *            <code>true</code> if the comment header for this data type
	 *            usually contains a version number
	 * @param parser
	 *            the line parser for transforming lines from this data type
	 *            into objects; may not be <code>null</code>
	 * @param hints
	 *            a collection of resource name hints for identifying the
	 *            resource that contains the data. May be <code>null</code>, but
	 *            may not contain <code>null</code>
	 * @throws NullPointerException
	 *             if the specified parser is <code>null</code>
	 * @since JWI 2.0.0
	 */
    public DataType(String userFriendlyName, boolean hasVersion, ILineParser<T> parser, Collection<String> hints) {
    	if(parser == null) 
    		throw new NullPointerException();
        this.name = userFriendlyName;
        this.parser = parser;
        this.hasVersion = hasVersion;
        this.hints = (hints == null || hints.isEmpty()) ? Collections.<String>emptySet() : Collections.unmodifiableSet(new HashSet<String>(hints));
    }
    
    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.data.IDataType#hasVersion()
     */
    public boolean hasVersion(){
    	return hasVersion;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.data.IDataType#getResourceNameHints()
     */
    public Set<String> getResourceNameHints() {
        return hints;
    }
    
    /* 
     * (non-Javadoc) 
     *
     * @see edu.mit.jwi.data.IDataType#getParser()
     */
    public ILineParser<T> getParser(){
    	return parser;
    }

    /* 
     * (non-Javadoc) 
     *
     * @see java.lang.Object#toString()
     */
    public String toString() {
        return name;
    }
    
	// set of all data types implemented in this class
	private static Set<DataType<?>> dataTypes = null;
	
	/**
	 * Emulates the Enum.values() function.
	 * 
	 * @return all the static data type instances listed in the class, in the
	 *         order they are declared.
	 * @since JWI 2.0.0
	 */
	public static Collection<DataType<?>> values(){
		if(dataTypes == null){
			
			// get all the fields containing ContentType
			Field[] fields = DataType.class.getFields();
			List<Field> instanceFields = new ArrayList<Field>(); 
			for(Field field : fields)
				if(field.getGenericType() == DataType.class)
					instanceFields.add(field);
			
			// this is the backing set
			Set<DataType<?>> hidden = new LinkedHashSet<DataType<?>>(instanceFields.size());
			
			// fill in the backing set
			DataType<?> dataType;
			for(Field field : instanceFields){
				try{
					dataType = (DataType<?>)field.get(null);
					if(dataType != null) 
						hidden.add(dataType);
				} catch(IllegalAccessException e){
					// Ignore
				}
			}
			
			// make the value set unmodifiable
			dataTypes = Collections.unmodifiableSet(hidden);
		}
			
		return dataTypes;
	}

	/**
	 * Finds the first file that satisfies the naming constraints of both
	 * the data type and part of speech.
	 * 
	 * @param type
	 *            the data type whose resource name hints should be used, may
	 *            not be <code>null</code>
	 * @param pos
	 *            the part of speech whose resource name hints should be used,
	 *            may be <code>null</code>
	 * @param files
	 *            the files to be search, may be empty but not <code>null</code>
	 * @return the file that matches both the pos and type naming conventions,
	 *         or <code>null</code> if none is found.
	 * @throws NullPointerException
	 *             if the data type or file collection is <code>null</code>
	 * @since JWI 2.2.0
	 */
	public static File find(IDataType<?> type, POS pos, Collection<? extends File> files){

		Set<String> typePatterns = type.getResourceNameHints();
		Set<String> posPatterns = (pos == null) ? Collections.<String>emptySet() : pos.getResourceNameHints();
		
		String name;
		for (File file : files) {
			name = file.getName().toLowerCase(); // added toLowerCase() as fix for Bug 017
			if(containsOneOf(name, typePatterns) && containsOneOf(name, posPatterns)) 
				return file;
		}
		
		return null;
	}
	
	
	/**
	 * Finds the first file that satisfies the naming constraints of both
	 * the data type and part of speech.
	 * 
	 * @since JWI 2.3.3-hadoop
	 */
	public static FileStatus find(IDataType<?> type, POS pos, List<FileStatus> files){

		Set<String> typePatterns = type.getResourceNameHints();
		Set<String> posPatterns = (pos == null) ? Collections.<String>emptySet() : pos.getResourceNameHints();
		
		String name;
		for (FileStatus file : files) {
			name = file.getPath().getName().toLowerCase(); // added toLowerCase() as fix for Bug 017
			if(containsOneOf(name, typePatterns) && containsOneOf(name, posPatterns)) 
				return file;
		}
		
		return null;
	}

	/**
	 * Checks to see if one of the string patterns specified in the set of
	 * strings is found in the specified target string. If the pattern set is
	 * empty or null, returns <code>true</code>. If a pattern is found in the
	 * target string, returns <code>true</code>. Otherwise, returns
	 * <code>false</code>.
	 * 
	 * @param target
	 *            the string to be searched
	 * @param patterns
	 *            the patterns to search for
	 * @return <code>true</code> if the target contains one of the patterns;
	 *         <code>false</code> otherwise
	 * @since JWI 2.2.0
	 */
	public static boolean containsOneOf(String target, Set<String> patterns) {
		if (patterns == null || patterns.size() == 0) 
			return true;
		for (String pattern : patterns)
			if (target.indexOf(pattern) > -1) 
				return true;
		return false;
	}

}